third_party/proton/csrc/include/Profiler/CuptiProfiler.h

```diff
@@ -2,12 +2,13 @@
 #define PROTON_PROFILER_CUPTI_PROFILER_H_
 
 #include "Context/Context.h"
-#include "Driver/GPU/Cupti.h"
 #include "Profiler.h"
 
 #include <atomic>
 #include <map>
 
+struct CUctx_st;
```

Comment:
Explain why we forward-declare instead of including the header (basically this commit message)

Reply:
Indeed it seems very important to explain that this header cannot include nvidia-specific stuff but the cpp file can.

You could consider using the PIMPL pattern?

Reply:
> Oh, I see. This file is indeed specific to cupti, but it's included by the generic session handler.

Yeah, this is the reason why we need forward declaration. 

Can you explain a bit about how PIMPL could be used here to solve the problem?

Reply:
With PIMPL you wouldn't have any private member functions in the header.  It's the private member function that's requiring the cupti forward-declaration.  (You don't, and can't, have any cupti types in the public signature of the class.)

```
// cupti_profiler.h

class CuptiProfiler {
public:
  int foo() { return pimpl->foo(); }
  ...
private:
  CuptiProfilerPimpl* pimpl;
};

// cupti_profiler.cc
#include <cupti.h>

class CuptiProfilerPimpl { ... };
```

Reply:
Refined the commit message

Reply:
Oh, I see what you mean. Sure I can rework the implementation to use this pattern

Reply:
Hi @jlebar, code refactoring is done. Feel free to take a second look. Thanks for the suggestions.